{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preamble"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 0\n",
    "%autoreload"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import xarray as xr\n",
    "import sqlite3\n",
    "import pandas as pd\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "from lib.pandas_util import idxwhere\n",
    "import warnings\n",
    "import torch\n",
    "import pyro\n",
    "import scipy as sp\n",
    "\n",
    "\n",
    "import sfacts as sf\n",
    "from sfacts.model_zoo import (\n",
    "    full_metagenotype,\n",
    "    full_metagenotype_no_missing,\n",
    "    full_metagenotype_pp_pi,\n",
    "    full_metagenotype_dirichlet_rho,\n",
    "    full_metagenotype_dirichlet_rho_no_missing,\n",
    "    full_metagenotype_pp_pi_no_missing,\n",
    "#     full_metagenotype_special_meta,\n",
    "#     full_metagenotype_no_missing_special_meta,\n",
    ")\n",
    "\n",
    "from sklearn.cluster import AgglomerativeClustering\n",
    "\n",
    "\n",
    "from lib.project_style import color_palette, major_allele_frequency_bins\n",
    "from lib.project_data import metagenotype_db_to_xarray\n",
    "from lib.plot import ordination_plot, mds_ordination, nmds_ordination\n",
    "import lib.plot\n",
    "from lib.plot import construct_ordered_pallete"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "warnings.filterwarnings(\n",
    "    \"ignore\",\n",
    "    message=\"torch.tensor results are registered as constants in the trace. You can safely ignore this warning if you use this function to create tensors out of constant variables that would be the same every time you call this function. In any other case, this might cause the trace to be incorrect.\",\n",
    "    category=torch.jit.TracerWarning,\n",
    "#     module=\"trace_elbo\",  # FIXME: What is the correct regex for module?\n",
    "#     lineno=5,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mpl.rcParams['figure.dpi'] = 200"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Style"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set Focal Species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_of_interest = '104345'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load from DB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "con = sqlite3.connect('data/core.a.proc.gtpro.2.denorm.db')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_mgen_all = pd.read_sql(f\"\"\"\n",
    "SELECT\n",
    "  lib_id\n",
    ", species_position\n",
    ", reference_tally\n",
    ", alternative_tally\n",
    "FROM snp_x_lib\n",
    "JOIN lib USING (lib_id)\n",
    "WHERE species_id = '{species_of_interest}'\n",
    "  AND lib_type = 'metagenome'\n",
    "\"\"\", con=con, index_col=['lib_id', 'species_position'])\n",
    "\n",
    "mgen_all = sf.data.Metagenotypes(metagenotype_db_to_xarray(_mgen_all))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mgen_filt = (\n",
    "    mgen_all\n",
    "    .select_variable_positions(thresh=0.05)\n",
    "    .select_samples_with_coverage(0.05)\n",
    ")\n",
    "mgen_ss = mgen_filt.random_sample(position=1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mgen_filt.sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_drplt_all = pd.read_sql(f\"\"\"\n",
    "SELECT\n",
    "  lib_id\n",
    ", species_position\n",
    ", reference_tally\n",
    ", alternative_tally\n",
    "FROM snp_x_lib\n",
    "JOIN lib USING (lib_id)\n",
    "WHERE species_id = '{species_of_interest}'\n",
    "  AND lib_type = 'droplet'\n",
    "\"\"\", con=con, index_col=['lib_id', 'species_position'])\n",
    "\n",
    "drplt_all = sf.data.Metagenotypes(metagenotype_db_to_xarray(_drplt_all))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "drplt_to_sample = pd.read_sql(\"SELECT lib_id, sample_id FROM lib WHERE lib_type = 'droplet'\", con=con, index_col='lib_id').squeeze()\n",
    "drplt_to_sample.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "drplt_filt = drplt_all.mlift('reindex', position=mgen_filt.position, fill_value=0).select_samples_with_coverage(0.01)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "drplt_filt.sizes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Depth duplication across droplets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nposition = min(mgen_filt.sizes['position'], 1000)\n",
    "\n",
    "d = (\n",
    "        sf.data.Metagenotypes.concat(\n",
    "        dict(\n",
    "            mgen=mgen_filt,\n",
    "            drplt=drplt_filt,\n",
    "        ),\n",
    "        dim='sample'\n",
    "    )\n",
    "    .select_samples_with_coverage(0.01)\n",
    "    .random_sample(position=nposition)\n",
    ")\n",
    "\n",
    "sf.plot.plot_depth(\n",
    "    d.to_world(),\n",
    "    col_colors_func=lambda w: xr.Dataset(dict(\n",
    "        focal=w.sample.isin(['mgen_SS01009.m', 'mgen_SS01057.m']),\n",
    "        drplt=w.sample.str.startswith('drplt'),\n",
    "        depth=w.metagenotypes.mean_depth().pipe(np.log),\n",
    "        mgen_entropy=w.metagenotypes.entropy('sample'),\n",
    "    )),\n",
    ")\n",
    "\n",
    "sf.plot.plot_metagenotype(\n",
    "    d.to_world(),\n",
    "    col_colors_func=lambda w: xr.Dataset(dict(\n",
    "        focal=w.sample.isin(['mgen_SS01009.m', 'mgen_SS01057.m']),\n",
    "        drplt=w.sample.str.startswith('drplt'),\n",
    "        depth=w.metagenotypes.mean_depth().pipe(np.log),\n",
    "        mgen_entropy=w.metagenotypes.entropy('sample'),\n",
    "    )),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "distance_threshold = 0.5\n",
    "\n",
    "sf.plot.plot_metagenotype(\n",
    "    drplt_filt.to_world(),\n",
    "    col_colors_func=lambda d: pd.Series(\n",
    "        (\n",
    "            AgglomerativeClustering(\n",
    "                affinity='cosine',\n",
    "                distance_threshold=distance_threshold,\n",
    "                n_clusters=None,\n",
    "                linkage='complete',\n",
    "            )\n",
    "            .fit(d.metagenotypes.total_counts())\n",
    "            .labels_\n",
    "        ),\n",
    "        d.sample,\n",
    "        name='clust',\n",
    "    ).to_xarray(),\n",
    "#     col_linkage_func=lambda w: w.metagenotypes.linkage(pseudo=1.),\n",
    "    row_col_annotation_cmap=mpl.cm.hsv,\n",
    "    isel=dict(position=slice(0, 1000))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = drplt_filt\n",
    "\n",
    "drplt_agg = pd.Series(\n",
    "    (\n",
    "        AgglomerativeClustering(\n",
    "            affinity='cosine',\n",
    "            distance_threshold=distance_threshold,\n",
    "            n_clusters=None,\n",
    "            linkage='complete'\n",
    "        )\n",
    "        .fit(d.total_counts())\n",
    "        .labels_\n",
    "    ),\n",
    "    d.sample,\n",
    "    name='clust',\n",
    ")\n",
    "\n",
    "drplt_agg_pal = construct_ordered_pallete(drplt_agg, cm='hsv')\n",
    "\n",
    "ordination_plot(\n",
    "    d.pdist(),\n",
    "    ordin=lib.plot.nmds_ordination,\n",
    "    meta=drplt_agg.to_frame(),\n",
    "    colorby='clust',\n",
    "    color_palette=drplt_agg_pal,\n",
    "    ordin_kws={'is_dmat': True},\n",
    "    fill_legend=False,\n",
    "    scatter_kws=dict(lw=0),\n",
    ")\n",
    "None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = drplt_filt\n",
    "\n",
    "ordination_plot(\n",
    "    d.pdist(),\n",
    "    ordin=lib.plot.tsne_ordination,\n",
    "    meta=drplt_agg.to_frame(),\n",
    "    colorby='clust',\n",
    "    color_palette=drplt_agg_pal,\n",
    "    ordin_kws={'is_dmat': True},\n",
    "    fill_legend=False,\n",
    "    scatter_kws=dict(lw=0),\n",
    ")\n",
    "None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "drplt_filt_derep = sf.data.Metagenotypes(\n",
    "    drplt_filt\n",
    "    .to_series()\n",
    "    .reset_index()\n",
    "    .rename(columns={0: 'tally'})\n",
    "    .join(drplt_agg, on='sample')\n",
    "    .join(drplt_to_sample, on='sample')\n",
    "    .groupby(['sample_id', 'clust', 'position', 'allele'])\n",
    "    .sum()\n",
    "    .astype(int)\n",
    "    .reset_index()\n",
    "    .assign(label=lambda x: x.sample_id + '_' + x.clust.astype(str))\n",
    "    .set_index(['label', 'position', 'allele'])\n",
    "    .rename_axis(index={'label': 'sample'})\n",
    "    .tally\n",
    "    .to_xarray()\n",
    ")\n",
    "print(drplt_filt.sizes, drplt_filt_derep.sizes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = drplt_filt_derep\n",
    "\n",
    "ax, *_ = ordination_plot(\n",
    "    d.pdist(),\n",
    "    ordin=lib.plot.nmds_ordination,\n",
    "    meta=d.sample.to_series().to_frame(),\n",
    "    ordin_kws={'is_dmat': True},\n",
    "    fill_legend=False,\n",
    "    scatter_kws=dict(lw=0),\n",
    ")\n",
    "\n",
    "ax.set_xlabel('PCo1')\n",
    "ax.set_ylabel('PCo2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "        sf.data.Metagenotypes.concat(\n",
    "        dict(\n",
    "            mgen=mgen_filt,\n",
    "            drplt=drplt_filt_derep,\n",
    "        ),\n",
    "        dim='sample'\n",
    "    )\n",
    "    .select_samples_with_coverage(0.001)\n",
    "    .random_sample(position=1000)\n",
    ")\n",
    "\n",
    "sf.plot.plot_depth(\n",
    "    d.to_world(),\n",
    "    col_colors_func=lambda w: xr.Dataset(dict(\n",
    "        focal=w.sample.isin(['mgen_SS01009.m', 'mgen_SS01057.m']),\n",
    "        drplt=w.sample.str.startswith('drplt'),\n",
    "        depth=w.metagenotypes.mean_depth().pipe(np.log),\n",
    "        mgen_entropy=w.metagenotypes.entropy('sample'),\n",
    "    )),\n",
    ")\n",
    "\n",
    "sf.plot.plot_metagenotype(\n",
    "    d.to_world(),\n",
    "    col_colors_func=lambda w: xr.Dataset(dict(\n",
    "        focal=w.sample.isin(['mgen_SS01009.m', 'mgen_SS01057.m']),\n",
    "        drplt=w.sample.str.startswith('drplt'),\n",
    "        depth=w.metagenotypes.mean_depth().pipe(np.log),\n",
    "        mgen_entropy=w.metagenotypes.entropy('sample'),\n",
    "    )),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "g = sf.plot.plot_metagenotype(\n",
    "    world=mgen_filt.select_samples_with_coverage(0.05).mlift('sel', position=mgen_ss.position).to_world(),\n",
    "    matrix_func=lambda w: w.metagenotypes.alt_allele_fraction(pseudo=0.0).T,\n",
    "    col_colors_func=None,\n",
    ")\n",
    "\n",
    "g.ax_heatmap.set_facecolor('darkgrey')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "g = sf.plot.plot_metagenotype(\n",
    "    world=mgen_filt.select_samples_with_coverage(0.05).mlift('sel', position=mgen_ss.position).to_world(),\n",
    "#     matrix_func=lambda w: w.metagenotypes.alt_allele_fraction(pseudo=0.0).T,\n",
    "    col_colors_func=None,\n",
    ")\n",
    "\n",
    "g.ax_heatmap.set_facecolor('darkgrey')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "\n",
    "sf.plot.plot_metagenotype_frequency_spectrum(mgen_filt.to_world(), sample_list=['SS01009.m'], axs=ax, bins=np.linspace(0.5, 1.0, num=51))\n",
    "# ax.set_yscale('log')\n",
    "ax.set_ylim(0, 400)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mgen_filt.mean_depth().sel(sample=['SS01009.m', 'SS01060.m', 'SS01033.m'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "\n",
    "sf.plot.plot_metagenotype_frequency_spectrum(mgen_filt.to_world(), sample_list=['SS01060.m'], axs=ax, bins=np.linspace(0.5, 1.0, num=51))\n",
    "# ax.set_yscale('log')\n",
    "ax.set_ylim(0, 500)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "\n",
    "sf.plot.plot_metagenotype_frequency_spectrum(mgen_filt.to_world(), sample_list=['SS01033.m'], axs=ax, bins=np.linspace(0.5, 1.0, num=51))\n",
    "# ax.set_yscale('log')\n",
    "ax.set_ylim(0, 500)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Haplotyping"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fitting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nstrain = 50\n",
    "nposition = min(int(3e4), mgen_filt.sizes['position'])\n",
    "\n",
    "est1 = (\n",
    "# est, _ = (\n",
    "    sf.workflow.fit_subsampled_metagenotype_collapse_strains_then_iteratively_refit_full_genotypes_no_missing(\n",
    "        full_metagenotype_no_missing.model_structure,\n",
    "#         mgen_ss,\n",
    "        mgen_filt.random_sample(position=nposition),\n",
    "        nstrain=nstrain,\n",
    "        hyperparameters=dict(\n",
    "            gamma_hyper=0.1,\n",
    "#             delta_hyper_r=0.8,\n",
    "#             delta_hyper_temp=0.01,\n",
    "            rho_hyper=1.0,\n",
    "            pi_hyper=1.0,\n",
    "            epsilon_hyper_mode=0.01,\n",
    "            epsilon_hyper_spread=1.5,\n",
    "            alpha_hyper_hyper_mean=1000.0,\n",
    "            alpha_hyper_hyper_scale=0.5,\n",
    "            alpha_hyper_scale=2.0,\n",
    "#             m_hyper_hyper_r_mean=10.,\n",
    "#             m_hyper_hyper_r_scale=1.,\n",
    "#             m_hyper_r_scale=1.,\n",
    "        ),\n",
    "        nposition=nposition,\n",
    "        diss_thresh=0.05,\n",
    "        frac_thresh=1e-3,\n",
    "        stage2_hyperparameters=dict(gamma_hyper=1.0),\n",
    "        device='cuda',\n",
    "        dtype=torch.float64,\n",
    "        quiet=False,\n",
    "        estimation_kwargs=dict(\n",
    "            lagA=10,\n",
    "            lagB=100,\n",
    "            opt=pyro.optim.Adamax({\"lr\": 1.0}, {\"clip_norm\": 100}),\n",
    "            seed=2,\n",
    "            jit=True,\n",
    "        )\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "e = est1\n",
    "\n",
    "g = sf.plot.plot_community(\n",
    "    e,\n",
    "    col_linkage_func=lambda w: w.communities.linkage(dim=\"sample\"),\n",
    "    row_linkage_func=lambda w: w.genotypes.linkage(dim=\"strain\"),\n",
    "#     norm=None,\n",
    "#     col_colors_func=lambda w: xr.Dataset(dict(\n",
    "#         focal=w.sample.isin(['SS01009.m', 'SS01057.m']),\n",
    "#         depth=w.metagenotypes.mean_depth().pipe(np.sqrt),\n",
    "#         m_ent=w.metagenotypes.entropy('sample'),\n",
    "#         c_ent=w.communities.entropy('sample'),\n",
    "#         alpha=w.data.alpha.pipe(np.log),\n",
    "#         epsilon=w.data.epsilon,\n",
    "#         mu=w.data.mu.pipe(np.sqrt),\n",
    "#     )),\n",
    "    col_colors_func=None,\n",
    "    row_colors_func=None,\n",
    "    norm=mpl.colors.PowerNorm(1/2),\n",
    "    pad_height=2.,\n",
    "#     cheight=0.2,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "e = est1\n",
    "\n",
    "sf.plot.plot_genotype(\n",
    "    e,\n",
    "    col_linkage_func=lambda w: w.genotypes.linkage(dim=\"position\"),\n",
    "    row_linkage_func=lambda w: w.genotypes.linkage(dim=\"strain\"),\n",
    "    row_colors_func=None,\n",
    "#     row_colors_func=lambda w: xr.Dataset(dict(\n",
    "#         depth=(w.communities.data * w.metagenotypes.mean_depth()).mean(\"sample\").pipe(np.log),\n",
    "# #         focal=(w.communities.data.sel(sample='SS01009.m')).pipe(np.cbrt),\n",
    "#         entropy=w.genotypes.entropy(),\n",
    "#     )),\n",
    "    transpose=True,\n",
    "    scalex=1e-3,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "e = est1\n",
    "\n",
    "g = sf.plot.plot_metagenotype(\n",
    "    e,\n",
    "    row_linkage_func=lambda w: w.genotypes.linkage(dim=\"position\"),\n",
    "    col_linkage_func=lambda w: w.communities.linkage(dim=\"sample\"),\n",
    "#     row_linkage_func=lambda w: w.genotypes.linkage(dim=\"strain\"),\n",
    "#     row_colors_func=lambda w: xr.Dataset(dict(\n",
    "#         depth=(w.communities.data * w.metagenotypes.mean_depth()).mean(\"sample\").pipe(np.log),\n",
    "# #         focal=(w.communities.data.sel(sample='SS01009.m')).pipe(np.cbrt),\n",
    "#         entropy=w.genotypes.entropy(),\n",
    "#     )),\n",
    "    col_colors_func=None,\n",
    "    scaley=1e-3,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "e = est1\n",
    "\n",
    "g = sf.plot.plot_metagenotype(\n",
    "    e,\n",
    "    matrix_func=lambda w: w.metagenotypes.alt_allele_fraction(pseudo=0.0).T,\n",
    "    row_linkage_func=lambda w: w.genotypes.linkage(dim=\"position\"),\n",
    "    col_linkage_func=lambda w: w.communities.linkage(dim=\"sample\"),\n",
    "#     row_linkage_func=lambda w: w.genotypes.linkage(dim=\"strain\"),\n",
    "#     row_colors_func=lambda w: xr.Dataset(dict(\n",
    "#         depth=(w.communities.data * w.metagenotypes.mean_depth()).mean(\"sample\").pipe(np.log),\n",
    "#         focal=(w.communities.data.sel(sample='SS01009.m')).pipe(np.cbrt),\n",
    "#         entropy=w.genotypes.entropy(),\n",
    "#     )),\n",
    "    col_colors_func=None,\n",
    "    scaley=1e-3,\n",
    ")\n",
    "\n",
    "g.ax_heatmap.set_facecolor('darkgrey')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# d = drplt_filt_derep.select_samples_with_coverage(0.1).data.\n",
    "\n",
    "drplt_strain_type = sf.estimation.strain_cluster(\n",
    "    (\n",
    "        drplt_filt_derep\n",
    "        .select_samples_with_coverage(0.05)\n",
    "        .to_estimated_genotypes(pseudo=1.0)\n",
    "        .to_world()\n",
    "    ),\n",
    "    thresh=0.1,\n",
    ")\n",
    "                     \n",
    "drplt_strain_type_palette = lib.plot.construct_ordered_pallete(sorted(drplt_strain_type.unique()), cm='Spectral')\n",
    "drplt_strain_type.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "w = sf.data.World(xr.Dataset(dict(\n",
    "    metagenotypes=drplt_filt_derep.select_samples_with_coverage(0.05).data,\n",
    "    genotypes=est1.genotypes.data\n",
    ")))\n",
    "\n",
    "g = sf.plot.plot_metagenotype(\n",
    "    world=w.sel(position=mgen_ss.position),\n",
    "    matrix_func=lambda w: w.metagenotypes.alt_allele_fraction(pseudo=0.0).T,\n",
    "    row_linkage_func=lambda w: w.genotypes.linkage('position'),\n",
    "    col_linkage_func=lambda w: w.metagenotypes.linkage('sample'),\n",
    "#     col_colors_func=lambda w: drplt_strain_type.loc[w.sample].to_xarray().rename('this').to_dataset(),\n",
    "    col_colors_func=None,\n",
    "    col_colors=drplt_strain_type.map(drplt_strain_type_palette),\n",
    "    scalex=0.2,\n",
    "#     scaley=2e-3,\n",
    "    tree_kws=dict(lw=2.),\n",
    "    dwidth=1e-5,\n",
    "    dheight=2.5,\n",
    ")\n",
    "\n",
    "g.ax_heatmap.set_facecolor('darkgrey')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "w = sf.data.World(xr.Dataset(dict(\n",
    "    metagenotypes=drplt_filt_derep.select_samples_with_coverage(0.05).data,\n",
    "    genotypes=est1.genotypes.data\n",
    ")))\n",
    "\n",
    "g = sf.plot.plot_metagenotype(\n",
    "    world=w.sel(position=mgen_ss.position),\n",
    "    row_linkage_func=lambda w: w.genotypes.linkage('position'),\n",
    "    col_linkage_func=lambda w: w.metagenotypes.linkage('sample'),\n",
    "#     col_colors_func=lambda w: drplt_strain_type.loc[w.sample].to_xarray().rename('this').to_dataset(),\n",
    "    col_colors_func=None,\n",
    "    col_colors=drplt_strain_type.map(drplt_strain_type_palette),\n",
    "    scalex=0.2,\n",
    "#     scaley=2e-3,\n",
    "    tree_kws=dict(lw=2.),\n",
    "    dwidth=1e-5,\n",
    "    dheight=2.5,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "w = sf.data.World(xr.Dataset(dict(\n",
    "    metagenotypes=drplt_filt_derep.select_samples_with_coverage(0.01).data,\n",
    "    genotypes=est1.genotypes.data\n",
    ")))\n",
    "\n",
    "g = sf.plot.plot_metagenotype(\n",
    "    world=w.isel(position=slice(0, 2000)),\n",
    "    matrix_func=lambda w: w.metagenotypes.alt_allele_fraction(pseudo=0.0).T,\n",
    "    row_linkage_func=lambda w: w.genotypes.linkage('position'),\n",
    "    col_linkage_func=lambda w: w.metagenotypes.linkage('sample'),\n",
    "#     col_colors_func=lambda w: drplt_strain_type.loc[w.sample].to_xarray().rename('this').to_dataset(),\n",
    "    col_colors_func=None,\n",
    "    scalex=0.2,\n",
    "#     scaley=2e-3,\n",
    "    tree_kws=dict(lw=2.),\n",
    "    dwidth=1e-5,\n",
    "    dheight=2.5,\n",
    "#     isel=dict(position=slice(0, 2000)),\n",
    ")\n",
    "\n",
    "g.ax_heatmap.set_facecolor('darkgrey')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.cluster.hierarchy import linkage\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "\n",
    "w = sf.data.World(xr.Dataset(dict(\n",
    "    metagenotypes=drplt_filt.select_samples_with_coverage(0.05).data,\n",
    "    genotypes=est1.genotypes.data\n",
    ")))\n",
    "\n",
    "def depth_linkage(w, method=\"complete\", optimal_ordering=False, **kwargs):\n",
    "    d = w.metagenotypes.total_counts().to_series().unstack('position')\n",
    "    cdmat = pdist(d, 'cosine')\n",
    "    return linkage(cdmat, method=method, optimal_ordering=optimal_ordering, **kwargs)\n",
    "    \n",
    "    \n",
    "\n",
    "g = sf.plot.plot_metagenotype(\n",
    "    world=w.sel(position=mgen_ss.position),\n",
    "#     matrix_func=lambda w: w.metagenotypes.alt_allele_fraction(pseudo=0.0).T,\n",
    "    row_linkage_func=lambda w: w.genotypes.linkage('position'),\n",
    "    col_linkage_func=depth_linkage,\n",
    "#     col_colors_func=lambda w: drplt_strain_type.loc[w.sample].to_xarray().rename('this').to_dataset(),\n",
    "    col_colors_func=None,\n",
    "    col_colors=drplt_agg.map(drplt_agg_pal),\n",
    "    scalex=0.2,\n",
    "#     scaley=2e-3,\n",
    "    tree_kws=dict(lw=2.),\n",
    "    dwidth=1e-5,\n",
    "    dheight=2.5,\n",
    ")\n",
    "\n",
    "g.ax_heatmap.set_facecolor('darkgrey')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "w = sf.data.World(xr.Dataset(dict(\n",
    "    metagenotypes=drplt_filt_derep.select_samples_with_coverage(0.05).data,\n",
    "    genotypes=est1.genotypes.data\n",
    ")))\n",
    "\n",
    "ax, *_ = ordination_plot(\n",
    "    w.metagenotypes.pdist(),\n",
    "    ordin=lib.plot.nmds_ordination,\n",
    "    meta=drplt_strain_type.to_frame(name='clust'),\n",
    "    colorby='clust',\n",
    "    color_palette=drplt_strain_type_palette,\n",
    "    markersize_palette={'__none__': 40},\n",
    "    ordin_kws={'is_dmat': True},\n",
    "    fill_legend=False,\n",
    "    scatter_kws=dict(lw=0.5, alpha=0.9),\n",
    ")\n",
    "\n",
    "ax.set_xlabel('PCo1')\n",
    "ax.set_ylabel('PCo2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "e = est1\n",
    "sample_list = [s for s in e.sample.values if s in ['SS01009.m']]\n",
    "\n",
    "d = e.communities.sel(sample=sample_list).to_series()\n",
    "strain_list1 = d[d > 0.01].reset_index().strain.unique()\n",
    "print(strain_list1)\n",
    "d.sort_values(ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "e = est1.sel(position=mgen_ss.position)\n",
    "\n",
    "if len(strain_list1) > 1:\n",
    "    sf.plot.plot_genotype(\n",
    "        e,\n",
    "        col_linkage_func=lambda w: w.genotypes.linkage('position'),\n",
    "        row_linkage_func=None,\n",
    "#         row_linkage_func=lambda w: w.genotypes.linkage('strain'),\n",
    "#         scalex=1e-3,\n",
    "        row_colors_func=None,\n",
    "        transpose=True,\n",
    "        dheight=1e-5,\n",
    "        dwidth=1e-5,\n",
    "#         isel=dict(strain=slice(0, 5)),\n",
    "        scaley=0.5,\n",
    "        tree_kws=dict(lw=2.),\n",
    "        isel=dict(strain=[5, 10, 12, 14]),\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "w = sf.data.Genotypes.concat(dict(\n",
    "        drplt=drplt_filt_derep.select_samples_with_coverage(0.05).to_estimated_genotypes(),\n",
    "        est=est1.genotypes.mlift('sel', strain=strain_list1),\n",
    "), dim='strain')\n",
    "\n",
    "ax, *_ = ordination_plot(\n",
    "    w.pdist(),\n",
    "    ordin=lib.plot.nmds_ordination,\n",
    "    meta=pd.DataFrame(dict(\n",
    "        clust=drplt_strain_type.rename(lambda s: 'drplt_' + s),\n",
    "        is_est=w.strain.str.startswith('est_').to_series(),\n",
    "    )).fillna(-1),\n",
    "    colorby='clust',\n",
    "    color_palette=drplt_strain_type_palette,\n",
    "    markerby='is_est',\n",
    "    marker_palette={True: '>', False: 'o'},\n",
    "    zorderby='is_est',\n",
    "    markersizeby='is_est',\n",
    "    markersize_palette={True: 0, False: 40},\n",
    "    ordin_kws={'is_dmat': True},\n",
    "    fill_legend=False,\n",
    "    scatter_kws=dict(lw=0.5, alpha=0.95),\n",
    ")\n",
    "\n",
    "ax.set_xlabel('PCo1')\n",
    "ax.set_ylabel('PCo2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "w = sf.data.Genotypes.concat(dict(\n",
    "        drplt=drplt_filt_derep.select_samples_with_coverage(0.05).to_estimated_genotypes(),\n",
    "        est=est1.genotypes.mlift('sel', strain=strain_list1),\n",
    "), dim='strain')\n",
    "\n",
    "ax, *_ = ordination_plot(\n",
    "    w.pdist(),\n",
    "    ordin=lib.plot.nmds_ordination,\n",
    "    meta=pd.DataFrame(dict(\n",
    "        clust=drplt_strain_type.rename(lambda s: 'drplt_' + s),\n",
    "        is_est=w.strain.str.startswith('est_').to_series(),\n",
    "    )).fillna(-1),\n",
    "    colorby='clust',\n",
    "    color_palette=drplt_strain_type_palette,\n",
    "    markerby='is_est',\n",
    "    marker_palette={True: '>', False: 'o'},\n",
    "    zorderby='is_est',\n",
    "    markersizeby='is_est',\n",
    "    markersize_palette={True: 60, False: 40},\n",
    "    ordin_kws={'is_dmat': True},\n",
    "    fill_legend=False,\n",
    "    scatter_kws=dict(lw=0.5, alpha=0.95),\n",
    ")\n",
    "\n",
    "ax.set_xlabel('PCo1')\n",
    "ax.set_ylabel('PCo2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "w = sf.data.Genotypes.concat(dict(\n",
    "        drplt=drplt_filt_derep.select_samples_with_coverage(0.05).to_estimated_genotypes(),\n",
    "        est=est1.genotypes.mlift('sel', strain=strain_list1),\n",
    "), dim='strain')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "a = drplt_filt_derep.to_estimated_genotypes()\n",
    "b = est1.genotypes.drop_sel(strain=[-1])  # .sel(strain=strain_list1)\n",
    "\n",
    "drplt_accuracy = pd.DataFrame(\n",
    "    sf.math.genotype_cdist(a.values, b.values),\n",
    "    index=a.strain,\n",
    "    columns=b.strain,\n",
    ").assign(\n",
    "    best_strain=lambda x: x.idxmin(1),\n",
    "    best_of_sample_strains=lambda x: x[strain_list1].idxmin(1),\n",
    "    min_dist=lambda x: x.min(1),\n",
    "    min_dist_sample_strains=lambda x: x[strain_list1].min(1),\n",
    "    cvrg=(drplt_filt_derep.total_counts() > 0).mean(\"position\"),\n",
    "    median_dist=lambda x: x.median(1),\n",
    ")\n",
    "pal = construct_ordered_pallete(drplt_accuracy.best_strain.to_list(), cm='Paired')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for best_strain, d0 in drplt_accuracy.groupby('best_strain'):\n",
    "    for best_of_sample_strains, d1 in d0.groupby('best_of_sample_strains'):\n",
    "        plt.scatter('cvrg', 'min_dist_sample_strains', c=[pal[best_strain]], data=d1, label='__none__', alpha=0.7, lw={True: 1, False: 0}[best_strain in strain_list1], edgecolor='k')\n",
    "        \n",
    "for best_strain in sorted(drplt_accuracy.best_strain.unique()):\n",
    "    plt.scatter([], [], c=[pal[best_strain]], label=best_strain, lw={True: 1, False: 0}[best_strain in strain_list1], edgecolor='k')\n",
    "\n",
    "# for strain in drplt_accuracy.best_strain.unique():\n",
    "#     label = str(strain)\n",
    "#     if strain in strain_list1:\n",
    "#         label += '*'\n",
    "#     plt.scatter([], [], c=[pal[strain]], label=label)\n",
    "plt.xlabel('Horizontal Coverage')\n",
    "plt.ylabel('Strain Dissimilarity')\n",
    "plt.ylim(0.0, 0.25)\n",
    "plt.legend(ncol=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for best_strain, d0 in drplt_accuracy.groupby('best_strain'):\n",
    "    for best_of_sample_strains, d1 in d0.groupby('best_of_sample_strains'):\n",
    "        plt.scatter('cvrg', 'median_dist', c=[pal[best_strain]], data=d1, label='__none__', alpha=0.7, lw={True: 1, False: 0}[best_strain in strain_list1], edgecolor='k')\n",
    "        \n",
    "for best_strain in sorted(drplt_accuracy.best_strain.unique()):\n",
    "    plt.scatter([], [], c=[pal[best_strain]], label=best_strain, lw={True: 1, False: 0}[best_strain in strain_list1], edgecolor='k')\n",
    "\n",
    "# for strain in drplt_accuracy.best_strain.unique():\n",
    "#     label = str(strain)\n",
    "#     if strain in strain_list1:\n",
    "#         label += '*'\n",
    "#     plt.scatter([], [], c=[pal[strain]], label=label)\n",
    "plt.xlabel('Horizontal Coverage')\n",
    "plt.ylabel('Strain Dissimilarity')\n",
    "plt.ylim(0.0, 0.25)\n",
    "plt.legend(ncol=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.spatial.distance import pdist, squareform\n",
    "\n",
    "x = np.linspace(0, 1, num=201).reshape((-1, 1))\n",
    "d = pd.DataFrame(\n",
    "    squareform(pdist(x, sf.math.genotype_dissimilarity)),\n",
    "    index=x.squeeze(),\n",
    "    columns=x.squeeze()\n",
    ")\n",
    "\n",
    "ax = sns.heatmap(d, xticklabels=len(x)//5, yticklabels=len(x)//5, cmap='gray_r')\n",
    "ax.invert_yaxis()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Evaluation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "e = est1\n",
    "sample_list = [s for s in e.sample.values if s in ['SS01057.m', 'SS01009.m']]\n",
    "\n",
    "d = sf.data.Genotypes.concat(\n",
    "        dict(\n",
    "        est=e.genotypes.mlift('sel', strain=strain_list1),\n",
    "        other=e.genotypes,\n",
    "        mgen=(\n",
    "            mgen_filt\n",
    "            .mlift('sel', sample=sample_list)\n",
    "            .mlift('sel', position=e.position)\n",
    "            .to_estimated_genotypes()\n",
    "        ),\n",
    "        drplt=drplt_filt_derep.mlift('sel', position=e.position).to_estimated_genotypes(),\n",
    "    ),\n",
    "    dim='strain'\n",
    ")\n",
    "\n",
    "sf.plot.plot_genotype(\n",
    "    d.to_world(),\n",
    "    row_colors_func=None,\n",
    "    col_colors_func=None,\n",
    "    scalex=1e-3,\n",
    "    transpose=True,\n",
    "#     row_linkage_func=lambda w: w.genotypes.linkage(),\n",
    "#     col_linkage_func=None,\n",
    "#     col_cluster=False,\n",
    "#     row_cluster=False,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "e = est1\n",
    "sample_list = [s for s in e.sample.values if s in ['SS01057.m', 'SS01009.m']]\n",
    "\n",
    "\n",
    "d = sf.data.Genotypes.concat(dict(\n",
    "    est=e.genotypes.mlift('sel', strain=strain_list1),\n",
    "    othere=e.genotypes,\n",
    "    mgen=mgen_filt.mlift('sel', sample=sample_list).mlift('sel', position=e.position).to_estimated_genotypes(),\n",
    "    otherm=mgen_filt.mlift('sel', position=e.position).to_estimated_genotypes(),\n",
    "    drplt=drplt_filt_derep.select_samples_with_coverage(0.05).mlift('sel', position=e.position).to_estimated_genotypes(),\n",
    "), dim='strain')\n",
    "\n",
    "def _assign_gtype(x):\n",
    "    return x.index.to_series().str.split('_').apply(lambda x: x[0])\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(5, 5))\n",
    "\n",
    "_, ordin, _ = ordination_plot(\n",
    "    d.pdist(),\n",
    "    meta=(\n",
    "        pd.DataFrame([], index=d.strain)\n",
    "        .assign(gtype=_assign_gtype)\n",
    "        .assign(z=lambda x: (x.gtype.isin(['est', 'mgen']))\n",
    "                .astype(int))\n",
    "    ),\n",
    "    ordin=nmds_ordination,\n",
    "    ordin_kws=dict(is_dmat=True),\n",
    "    colorby='gtype',\n",
    "    zorderby='z',\n",
    "#     edgestyleby='gtype',\n",
    "#     edgestyle_palette={'drplt': 'None', 'est': '--', 'mgen': ':', 'other': ''}\n",
    "    scatter_kws=dict(lw=0),\n",
    "    ax=ax,\n",
    "    \n",
    ")\n",
    "\n",
    "for name, d1 in ordin[ordin.gtype.isin(['est', 'mgen'])].iterrows():\n",
    "    ax.annotate(name, xy=d1[['PC1', 'PC2']].to_list())\n",
    "None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "e = est1\n",
    "sample_list = [s for s in e.sample.values if s in ['SS01057.m', 'SS01009.m']]\n",
    "\n",
    "\n",
    "d = sf.data.Genotypes.concat(dict(\n",
    "    est=e.genotypes.mlift('sel', strain=strain_list1),\n",
    "#     othere=e.genotypes,\n",
    "    mgen=mgen_filt.mlift('sel', sample=sample_list).mlift('sel', position=e.position).to_estimated_genotypes(),\n",
    "#     otherm=mgen_filt.mlift('sel', position=e.position).to_estimated_genotypes(),\n",
    "    drplt=drplt_filt_derep.select_samples_with_coverage(0.05).mlift('sel', position=e.position).to_estimated_genotypes(),\n",
    "), dim='strain')\n",
    "\n",
    "def _assign_gtype(x):\n",
    "    return x.index.to_series().str.split('_').apply(lambda x: x[0])\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(5, 5))\n",
    "\n",
    "_, ordin, _ = ordination_plot(\n",
    "    d.pdist(),\n",
    "    meta=(\n",
    "        pd.DataFrame([], index=d.strain)\n",
    "        .assign(gtype=_assign_gtype)\n",
    "        .assign(z=lambda x: (x.gtype.isin(['est', 'mgen']))\n",
    "                .astype(int))\n",
    "    ),\n",
    "    ordin=nmds_ordination,\n",
    "    ordin_kws=dict(is_dmat=True),\n",
    "    colorby='gtype',\n",
    "    zorderby='z',\n",
    "#     edgestyleby='gtype',\n",
    "#     edgestyle_palette={'drplt': 'None', 'est': '--', 'mgen': ':', 'other': ''}\n",
    "    scatter_kws=dict(lw=0),\n",
    "    ax=ax,\n",
    "    \n",
    ")\n",
    "\n",
    "for name, d1 in ordin[ordin.gtype.isin(['est', 'mgen'])].iterrows():\n",
    "    ax.annotate(name, xy=d1[['PC1', 'PC2']].to_list())\n",
    "None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "e = est1\n",
    "sample_list = [s for s in e.sample.values if s in ['SS01057.m', 'SS01009.m']]\n",
    "\n",
    "d1 = sf.data.Genotypes.concat(\n",
    "        dict(\n",
    "#         est=e.genotypes.mlift('sel', strain=strain_list1),\n",
    "        other=e.genotypes,\n",
    "#         mgen=(\n",
    "#             mgen_filt\n",
    "#             .mlift('sel', sample=sample_list)\n",
    "#             .mlift('sel', position=e.position)\n",
    "#             .to_estimated_genotypes()\n",
    "#         ),\n",
    "    ),\n",
    "    dim='strain'\n",
    ")\n",
    "\n",
    "d2 = sf.data.Genotypes.concat(\n",
    "        dict(\n",
    "        drplt=drplt_filt_derep.select_samples_with_coverage(0.05).mlift('sel', position=e.position).to_estimated_genotypes(),\n",
    "    ),\n",
    "    dim='strain'\n",
    ")\n",
    "\n",
    "diss = pd.DataFrame(sf.math.genotype_cdist(d1.values, d2.values), index=d1.strain, columns=d2.strain)\n",
    "\n",
    "\n",
    "sns.clustermap(diss, vmax=0.05, row_linkage=d1.linkage(), col_linkage=d2.linkage())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "e = est1\n",
    "sample_list = [s for s in e.sample.values if s in ['SS01057.m', 'SS01009.m']]\n",
    "\n",
    "d = sf.data.Genotypes.concat(\n",
    "        dict(\n",
    "#         est=e.genotypes.mlift('sel', strain=strain_list1),\n",
    "        haplo=e.genotypes,\n",
    "#         mgen=(\n",
    "#             mgen_filt\n",
    "#             .mlift('sel', sample=sample_list)\n",
    "#             .mlift('sel', position=e.position)\n",
    "#             .to_estimated_genotypes()\n",
    "#         ),\n",
    "        drplt=(\n",
    "            drplt_filt_derep\n",
    "            .mlift('sel', position=e.position)\n",
    "            .select_samples_with_coverage(0.05)\n",
    "            .to_estimated_genotypes()\n",
    "        ),\n",
    "    ),\n",
    "    dim='strain'\n",
    ")\n",
    "\n",
    "sf.plot.plot_genotype(\n",
    "    d.to_world(),\n",
    "    row_colors_func=lambda w: xr.Dataset(dict(\n",
    "        haplo=(\n",
    "            w.strain.str.startswith('haplo').astype(int) +\n",
    "            w.strain.isin([f'haplo_{i}' for i in strain_list1]).astype(int)\n",
    "        ),\n",
    "#         haplo=w.strain.str.startswith('haplo'),\n",
    "    )),\n",
    "    col_colors_func=None,\n",
    "    scalex=1e-3,\n",
    "    dwidth=5.0,\n",
    "    dheight=0.1,\n",
    "    cwidth=1.,\n",
    "    transpose=True,\n",
    "#     row_linkage_func=lambda w: w.genotypes.linkage(),\n",
    "#     col_linkage_func=None,\n",
    "#     col_cluster=False,\n",
    "#     row_cluster=False,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nstrain = 50\n",
    "nposition = min(int(3e4), mgen_filt.sizes['position'])\n",
    "\n",
    "d = (\n",
    "        sf.data.Metagenotypes.concat(\n",
    "        dict(\n",
    "            mgen=mgen_filt,\n",
    "            drplt=drplt_filt_derep,\n",
    "        ),\n",
    "        dim='sample'\n",
    "    )\n",
    "    .select_samples_with_coverage(0.05)\n",
    ")\n",
    "\n",
    "est2 = (\n",
    "    sf.workflow.fit_subsampled_metagenotype_collapse_strains_then_iteratively_refit_full_genotypes_no_missing(\n",
    "        full_metagenotype_no_missing.model_structure,\n",
    "        d.random_sample(position=nposition),\n",
    "        nstrain=nstrain,\n",
    "        hyperparameters=dict(\n",
    "            gamma_hyper=0.1,\n",
    "            rho_hyper=1.0,\n",
    "            pi_hyper=1.0,\n",
    "            epsilon_hyper_mode=0.01,\n",
    "            epsilon_hyper_spread=1.5,\n",
    "            alpha_hyper_hyper_mean=1000.0,\n",
    "            alpha_hyper_hyper_scale=0.5,\n",
    "            alpha_hyper_scale=2.0,\n",
    "        ),\n",
    "        nposition=nposition,\n",
    "        diss_thresh=0.05,\n",
    "        frac_thresh=1e-3,\n",
    "        stage2_hyperparameters=dict(gamma_hyper=1.0),\n",
    "        device='cuda',\n",
    "        dtype=torch.float64,\n",
    "        quiet=False,\n",
    "        estimation_kwargs=dict(\n",
    "            lagA=10,\n",
    "            lagB=100,\n",
    "            opt=pyro.optim.Adamax({\"lr\": 1.0}, {\"clip_norm\": 100}),\n",
    "            seed=2,\n",
    "            jit=True,\n",
    "        )\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "e = est2\n",
    "\n",
    "g = sf.plot.plot_community(\n",
    "    e,\n",
    "#     col_linkage_func=lambda w: w.metagenotypes.linkage(dim=\"sample\"),\n",
    "    row_linkage_func=lambda w: w.genotypes.linkage(dim=\"strain\"),\n",
    "#     norm=None,\n",
    "    col_colors_func=lambda w: xr.Dataset(dict(\n",
    "        focal=w.sample.isin(['mgen_SS01009.m', 'mgen_SS01057.m']),\n",
    "        drplt=w.sample.str.startswith('drplt_'),\n",
    "        depth=w.metagenotypes.mean_depth().pipe(np.sqrt),\n",
    "        m_ent=w.metagenotypes.entropy('sample'),\n",
    "        c_ent=w.communities.entropy('sample'),\n",
    "        alpha=w.data.alpha.pipe(np.log),\n",
    "        epsilon=w.data.epsilon,\n",
    "        mu=w.data.mu.pipe(np.sqrt),\n",
    "    )),\n",
    "    norm=mpl.colors.PowerNorm(1/2),\n",
    "    pad_height=2.,\n",
    "#     cheight=0.2,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "e = est2\n",
    "\n",
    "sf.plot.plot_genotype(\n",
    "    e,\n",
    "#     col_linkage_func=lambda w: w.metagenotypes.linkage(dim=\"position\"),\n",
    "#     row_linkage_func=lambda w: w.genotypes.linkage(dim=\"strain\"),\n",
    "    row_colors_func=lambda w: xr.Dataset(dict(\n",
    "        depth=(w.communities.data * w.metagenotypes.mean_depth()).mean(\"sample\").pipe(np.log),\n",
    "#         focal=(w.communities.data.sel(sample='SS01009.m')).pipe(np.cbrt),\n",
    "        entropy=w.genotypes.entropy(),\n",
    "    )),\n",
    "    scalex=1e-3,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "e = est2\n",
    "sample_list = [s for s in e.sample.values if s in ['mgen_SS01057.m', 'mgen_SS01009.m'] or s.startswith('drplt_')]\n",
    "\n",
    "d = e.sel(sample=sample_list)\n",
    "#strain_list1 = d[d > 0.01].reset_index().strain.unique()\n",
    "#print(strain_list1)\n",
    "\n",
    "sf.plot.plot_community(\n",
    "    d,\n",
    "    col_colors_func=lambda w: xr.Dataset(dict(\n",
    "        focal=w.sample.isin(['mgen_SS01009.m', 'mgen_SS01057.m']),\n",
    "#         drplt=w.sample.str.startswith('drplt_'),\n",
    "#         depth=w.metagenotypes.mean_depth().pipe(np.sqrt),\n",
    "#         m_ent=w.metagenotypes.entropy('sample'),\n",
    "#         c_ent=w.communities.entropy('sample'),\n",
    "#         alpha=w.data.alpha.pipe(np.log),\n",
    "#         epsilon=w.data.epsilon,\n",
    "#         mu=w.data.mu.pipe(np.sqrt),\n",
    "    )),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "e = est2\n",
    "sample_list = [s for s in e.sample.values if s in ['mgen_SS01057.m', 'mgen_SS01009.m'] or s.startswith('drplt_')]\n",
    "\n",
    "d = (\n",
    "        sf.data.Metagenotypes.concat(\n",
    "        dict(\n",
    "            mgen=mgen_filt,\n",
    "            drplt=drplt_filt_derep,\n",
    "        ),\n",
    "        dim='sample'\n",
    "    )\n",
    "    .select_samples_with_coverage(0.05)\n",
    "    .mlift('sel', sample=sample_list)\n",
    ")\n",
    "\n",
    "sf.plot.plot_metagenotype(\n",
    "    d.to_world(),\n",
    "    col_colors_func=lambda w: xr.Dataset(dict(\n",
    "        focal=w.sample.isin(['mgen_SS01009.m', 'mgen_SS01057.m']),\n",
    "#         drplt=w.sample.str.startswith('drplt_'),\n",
    "#         depth=w.metagenotypes.mean_depth().pipe(np.sqrt),\n",
    "#         m_ent=w.metagenotypes.entropy('sample'),\n",
    "#         c_ent=w.communities.entropy('sample'),\n",
    "#         alpha=w.data.alpha.pipe(np.log),\n",
    "#         epsilon=w.data.epsilon,\n",
    "#         mu=w.data.mu.pipe(np.sqrt),\n",
    "    )),\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}